File: packages/db-schema/src/schema/calendar-sources.ts
import { sql } from 'drizzle-orm'
import { integer, sqliteTable, text, index, uniqueIndex } from 'drizzle-orm/sqlite-core'
import type { VectorClock } from '@memry/contracts/sync-api'

export type CalendarSourceKind = 'account' | 'calendar'
export type CalendarSourceSyncStatus = 'idle' | 'ok' | 'error' | 'pending'

export const calendarSources = sqliteTable(
  'calendar_sources',
  {
    id: text('id').primaryKey(),
    provider: text('provider').notNull(),
    kind: text('kind').$type<CalendarSourceKind>().notNull(),
    accountId: text('account_id'),
    remoteId: text('remote_id').notNull(),
    title: text('title').notNull(),
    timezone: text('timezone'),
    color: text('color'),
    isPrimary: integer('is_primary', { mode: 'boolean' }).notNull().default(false),
    isSelected: integer('is_selected', { mode: 'boolean' }).notNull().default(false),
    isMemryManaged: integer('is_memry_managed', { mode: 'boolean' }).notNull().default(false),
    syncCursor: text('sync_cursor'),
    syncStatus: text('sync_status').$type<CalendarSourceSyncStatus>().notNull().default('idle'),
    lastSyncedAt: text('last_synced_at'),
    metadata: text('metadata', { mode: 'json' }).$type<Record<string, unknown> | null>(),
    archivedAt: text('archived_at'),
    clock: text('clock', { mode: 'json' }).$type<VectorClock>(),
    syncedAt: text('synced_at'),
    createdAt: text('created_at')
      .notNull()
      .default(sql`(strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))`),
    modifiedAt: text('modified_at')
      .notNull()
      .default(sql`(strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))`)
  },
  (table) => [
    uniqueIndex('idx_calendar_sources_provider_remote').on(table.provider, table.kind, table.remoteId),
    index('idx_calendar_sources_account').on(table.accountId),
    index('idx_calendar_sources_selected').on(table.isSelected)
  ]
)

export type CalendarSource = typeof calendarSources.$inferSelect
export type NewCalendarSource = typeof calendarSources.$inferInsert

File: apps/desktop-tauri/src-tauri/src/db/calendar_sources.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "camelCase")]
pub struct CalendarSource {
    pub id: String,
    pub provider: String,
    pub kind: String,
    pub account_id: Option<String>,
    pub remote_id: String,
    pub title: String,
    pub timezone: Option<String>,
    pub color: Option<String>,
    pub is_primary: bool,
    pub is_selected: bool,
    pub is_memry_managed: bool,
    pub sync_cursor: Option<String>,
    pub sync_status: String,
    pub last_synced_at: Option<String>,
    pub metadata: Option<String>,
    pub archived_at: Option<String>,
    pub clock: Option<String>,
    pub synced_at: Option<String>,
    pub created_at: String,
    pub modified_at: String,
    pub last_error: Option<String>,
}

impl CalendarSource {
    pub fn from_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Self> {
        Ok(Self {
            id: row.get("id")?,
            provider: row.get("provider")?,
            kind: row.get("kind")?,
            account_id: row.get("account_id")?,
            remote_id: row.get("remote_id")?,
            title: row.get("title")?,
            timezone: row.get("timezone")?,
            color: row.get("color")?,
            is_primary: row.get::<_, i64>("is_primary")? != 0,
            is_selected: row.get::<_, i64>("is_selected")? != 0,
            is_memry_managed: row.get::<_, i64>("is_memry_managed")? != 0,
            sync_cursor: row.get("sync_cursor")?,
            sync_status: row.get("sync_status")?,
            last_synced_at: row.get("last_synced_at")?,
            metadata: row.get("metadata")?,
            archived_at: row.get("archived_at")?,
            clock: row.get("clock")?,
            synced_at: row.get("synced_at")?,
            created_at: row.get("created_at")?,
            modified_at: row.get("modified_at")?,
            last_error: row.get("last_error")?,
        })
    }
}
